### DISCUSSION TITLE
How do I get this extension to work with Theia?

### DISCUSSION THREAD

**Discussion by Qichar** [scalameta/metals-vscode#1244] [Q&A]
Hello, I am trying to get the scalameta.metals vscode extension to work with Theia (the browser-based vscode like IDE).  I started by downloading: 

scalameta.metals-1.20.0.vsix

and simply dropping it in the plugins folder.  It was recognized by Theia, and when I tried to open a .sc file it attempted to activate the extension.  However, I got this error:

root ERROR [hosted-plugin: 4807] Activating extension 'Scala (Metals)' failed: TypeError: Class extends value undefined is not a constructor or null
    at Object.<anonymous> (/tmp/vscode-unpacked/scalameta.metals-1.20.0.vsix/extension/node_modules/vscode-languageclient/lib/common/protocolTypeHierarchyItem.js:8:46)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at Function.module._load (/home/komchang/cse/node_modules/@theia/plugin-ext-vscode/lib/node/plugin-vscode-init.js:68:33)
    at Module.require (node:internal/modules/cjs/loader:1057:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/tmp/vscode-unpacked/scalameta.metals-1.20.0.vsix/extension/node_modules/vscode-languageclient/lib/common/codeConverter.js:18:37)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    
I was able to install this extension for use with vscode on my Mac just fine by following the instructions in the Readme.md file.  However, it's not working with Theia.  Does the "installer" do something special?  Every other vscode language extension has worked in Theia for me simply by dropping it into the plugins folder.  Does this extension have special setup requirements?

I am using node version 16.18.1.   I am using the extension provided by the "releases" section of this github, not one I built myself.

**Comment by tgodzik** [comment]
Thanks for the question! The last time I tried to use Eclipse Theia it was working mostly alright. What seems weird here is that the error is not in actual code we provide but in the library one. My guess is that we use newer version of vscode-languageclient that is not yet supported by theia?

**Reply by Qichar** [reply]
Thank you so much for the amazingly fast reply!  I am very grateful.

I am thinking the same thing you are thinking, and so I've been trying previous versions of your extension.  1.19.0 gives the same error as 1.20.0.  

I tried 1.10.10 and it gave a different error:

root ERROR [hosted-plugin: 10288] Activating extension 'Scala (Metals)' failed: Error: Dependent extension 'scala-lang.scala' is not installed.
    at /home/komchang/cse/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:208:39
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async PluginManagerExtImpl.$activatePlugin (/home/komchang/cse/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:271:13)
    at async Promise.all (index 0)
    at async PluginManagerExtImpl.activateBySingleEvent (/home/komchang/cse/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:266:9)
    at async PluginManagerExtImpl.$activateByEvent (/home/komchang/cse/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:246:13)
    
Maybe a year ago the scala-lang.scala extension was not bundled and it was a separate extension?  What was the latest version of the scalameta.metals extension where this was included and not separate?

I will keep trying different versions.  But of course I would prefer to use your latest one.  I tried looking up this error on the internet:
"Class extends value undefined is not a constructor or null"

It seems in javaScript this error is the result of a circular reference somewhere.  I have no idea if that is accurate, but it seems plausible.  I note that with 1.20.0 you are using vscode-languageclient version 8.0.2 which depends upon vscode-languageserver-protocol version 3.17.2.  However, in my version of Theia, I am using vscode-languageserver-protocol version 3.15.3 which is from 3 years ago (!), so you may very well be correct in your guess.

**Reply by tgodzik** [reply]
I would try an update the theia version and see if it fixes it. Out of curiosity, were some other extensions you installed also language ones?

**Reply by Qichar** [reply]
We have installed several language extensions and they worked fine with Theia, including java, python, go and csharp.  Note that to test your extension, I removed all the other plugins except yours.

I have gotten a previous version of your extension up and running in Theia.  This combination worked for me:
scalameta.metals-1.10.15.vsix
scala-lang.scala-0.5.6.vsix

When I tried moving to scalameta.metals-1.11.0.vsix, I ran into an issue where Theia does not support the Test Explorer API from Microsoft, and the extension would not activate at all.  I'm not sure if a later version of Theia implements the Test Explorer API, but I did find this thread from about a year ago:

https://github.com/eclipse-theia/theia/issues/10140

"As the testing API seems quite extensive, I can't imagine it being supported in the near future of Theia. Especially considering that we still have to do a lot of catching-up until we reach 1.59."

**Reply by tgodzik** [reply]
>and the extension would not activate at all.

Do you have a stack trace for the issue? Maybe it's something that we can solve?

Does it help if you change the configuration to use code lenses?

**Reply by tgodzik** [reply]
```
"metals.testUserInterface": "Code Lenses"
```

**Reply by Qichar** [reply]
Hi, thanks again for your help!

The error that prevents activation of 1.11.0 is:

root ERROR [hosted-plugin: 17434] Activating extension 'Scala (Metals)' failed: TypeError: Cannot read properties of undefined (reading 'createTestController')
    at new TestManager (/tmp/vscode-unpacked/scalameta.metals-1.12.0.vsix/extension/out/test-explorer/test-manager.js:45:46)
    at createTestManager (/tmp/vscode-unpacked/scalameta.metals-1.12.0.vsix/extension/out/test-explorer/test-manager.js:39:12)
    at /tmp/vscode-unpacked/scalameta.metals-1.12.0.vsix/extension/out/extension.js:412:66

**Reply by tgodzik** [reply]
Let's try with https://github.com/scalameta/metals-vscode/pull/1245

You can test out the change by doing `yarn build` and then using the resulting vsix.

**Reply by tgodzik** [reply]
We just merged it, could test it by downloading the latest vsix from https://marketplace.visualstudio.com/items?itemName=scalameta.metals ?
